packages/core/src/renderer/Color.ts

```diff
@@ -0,0 +1,715 @@
+};
+
+/**
+ * Creates a KNN Graph (adjacency list) where two nodes are connected
+ * if they are within each other's k nearest neighbors
+ * @param distGraph distance graph, created by shapeListToDistanceGraph
+ * @param k number of nearest neighbors, assumed k > 0 (k<=0 causes problems)
+ */
+const distGraphToKNNGraph = (distGraph: Graph, k: number): Ajlist => {
+  if (k >= distGraph.length) {
+    throw new Error("Warning: more neighbors requested than graph elems");
+  }
+
+  // ajlist will store the k closest neighbors for each node, in sorted order
+  var ajlist: Ajlist;
+
+  // first initialize the matrix
+  ajlist = createMatrix(distGraph.length, k);
+
+  // fill up the adjacency list
+  for (var i = 0; i < distGraph.length; i++) {
+    // list of all the neighbors of node i (includes itself)
+    var currRow = distGraph[i];
+
+    //  record the index of each neighbor
+    //  along with its distance away from node i
+    var indexedRow = currRow.map((element, index) => {
+      return [element, index];
+    });
+
+    // remove node i from the list
+    // (we don't include the node itself in its list of nbors)
+    indexedRow = indexedRow.filter((elemIndexPair) => {
+      return i !== elemIndexPair[1];
+    });
+
+    // sort the list based on the distance each node is away from node i
+    indexedRow.sort((e1, e2): number => {
+      return e1[0] - e2[0];
+    });
+
+    // put the indexes of the closest k neighbors in the ajlist
+    for (var j = 0; j < k; j++) {
+      ajlist[i][j] = indexedRow[j][1]; // add the indexes in sorted order
+    }
+  }
+
+  return ajlist;
+};
+
+// takes in a KNN graph (adjacency list)
+// creates a colorList, i.e. a list that maps node --> its assigned color
+/**
+ * Creates a colorlist c, where for every node i,
+ * c[i] === the color assigned to node i
+ * @param KNNGraph k-nearest neighbor graph (adjacency list),
+ *  created by distGraphToKNNGraph
+ * @param k number of nearest neighbors, assumed k > 0 (k<=0 causes problems)
+ * @param palette optional param, a list of colors to choose from
+ */
+const KNNGraphToColorList = (
+  KNNGraph: Ajlist,
+  k: number,
+  palette: RGB[] = viridis_data
+): RGB[] => {
+  // what's the minimum number of colors needed to color a graph, given k neighbors?
+  const numColorsRequested = 2 * k; // this is sufficient?
+
+  // list of colors that can be assigned to each node
+  const colorsToAssign = sampleUniformPalette(numColorsRequested, palette);
+
+  // initialize a colorlist
+  var colorList: RGB[] = [];
+  for (var i = 0; i < KNNGraph.length; i++) {
+    colorList.push([-1, -1, -1]);
+  }
+
+  // now assign the colors to the nodes of the graph in a greedy fashion.
+
+  // for every node...
+  for (var node = 0; node < KNNGraph.length; node++) {
+    // get its k nearest nbors...
+    const nodeNbors = KNNGraph[node];
+
+    var colorsThatCannotBeUsed: RGB[] = [];
+    // loop through its nbors, checking if any have already been assigned a color
```

Comment:
L220 - L236 can be done with a one-liner: `const bannedColors = nodeNbors.filter(n => n < node);`?

Reply:
There are probably other instances of this pattern. Lmk if you want me to do a closer pass of coding style